src/Interpreters/HashJoin.cpp

```diff
@@ -2294,14 +2308,52 @@ void HashJoin::joinBlockImplCross(Block & block, ExtraBlockPtr & not_processed)
                 const IColumn & column_right = *block_right.getByPosition(col_num).column;
                 dst_columns[num_existing_columns + col_num]->insertRangeFrom(column_right, 0, rows_right);
             }
+        };
+
+        for (const Block & compressed_block_right : data->blocks)
+        {
+            ++block_number;
+            if (block_number < start_right_block)
+                continue;
+
+            auto block_right = compressed_block_right.decompress();
+            process_right_block(block_right);
+            if (rows_added > max_joined_block_rows)
+            {
+                break;
+            }
+        }
+
+        if (tmp_stream && rows_added <= max_joined_block_rows)
+        {
+            if (reader == nullptr)
+            {
+                tmp_stream->finishWritingAsyncSafe();
+                reader = tmp_stream->getReadStream();
+            }
+            while (auto block_right = reader->read())
+            {
+                ++block_number;
+                process_right_block(block_right);
+                if (rows_added > max_joined_block_rows)
```

Comment:
By the way, I noticed that this setting is not randomized in out test.

Let's add it to `clickhouse-test`, so different scenarios are covered

```python
        "max_joined_block_size_rows": lambda: random.randint(8000, 100000),
```


https://github.com/ClickHouse/ClickHouse/blob/67cf40810a9fb96cfcd9119b8ba040471e017225/tests/clickhouse-test#L753